z3c/dependencychecker/dotted_name.py

```diff
@@ -0,0 +1,77 @@
+# -*- coding: utf-8 -*-
+import hashlib
+from functools import total_ordering
+
+
+@total_ordering
+class DottedName(object):
+
+    def __init__(
+        self,
+        name,
+        file_path=None,
+        line_number=None,
+        is_test=False,
+    ):
+        self.name = name
+        self.safe_name = self._get_safe_name(name)
+        self.namespaces = self._get_namespaces(self.safe_name)
+        self.is_namespaced = self._is_namespaced(self.namespaces)
+        self.file_path = file_path
+        self.line_number = line_number
+        self.is_test = is_test
+
+    @classmethod
+    def from_requirement(cls, requirement, file_path=None):
+        """A requirement in this method's context is a
+        pkg_resources.Requirement
+        """
+        return cls(
+            requirement.project_name,
+            file_path=file_path,
+        )
+
+    @staticmethod
+    def _get_safe_name(name):
+        safe_name = name.lower().replace('-', '_')
+        return safe_name
+
+    @staticmethod
+    def _get_namespaces(safe_name):
+        parts = safe_name.split('.')
+        return parts
+
+    @staticmethod
+    def _is_namespaced(namespaces):
+        return bool(len(namespaces) - 1)
+
+    def __lt__(self, other):
+        if not isinstance(other, DottedName):
+            return NotImplemented
+        return self.name < other.name
+
+    def __eq__(self, other):
+        if not isinstance(other, DottedName):
+            return NotImplemented
+
+        return self.safe_name == other.safe_name
```

Comment:
For equality you compare the safe_name, but for 'lighter than' you compare the name. That seems wrong: a dottedname of `Plone` would then be equal to `plone` (lowercase they are the same), but at the same time `Plone` would be lighter than `plone`.
I don't know if that is a problem in practice.
I would suggest changing it, unless you have a good reason not to (maybe the code was already like this).

Reply:
Given the test package used for testing this is how I had to put them to make it work.

If memory serves me well, the equality is to handle pypi mixed case and imports being different case and the sorting is due to alphabetically sorting the output:

https://github.com/reinout/z3c.dependencychecker/blob/master/z3c/dependencychecker/tests/test_integration.py#L16-L27

If we used safe_name then Products.GenericSetup would not be at the top but rather on the middle. I actually made the change and then pytest told me to undo it 🙂

Reply:
@mauritsvanrees although I merged I still want to hear back from you if you are ok with my explanation

Reply:
I would not mind having `Products.GenericSetup` end up in the middle. That makes it fit the case insensitive alphabetic sorting that we prefer in Plone.
I just worry that somewhere down the line it might give problems if `a==b ` and `a<b` are both true.
If it works for us, it may be no problem.

Reply:
I will keep that in mind, but so far it works just like the current implementation.

Ao although we could change that now, I would like to:
- have all #54 merged (in the small pull requests that I'm sending)
- create a 2.0.dev0 release and try it in the wild
- keep ironing out corner cases that have been missed
- make a final 2.0 release 🙂

Reply:
Fine with me.